Add support for mypy v1.16

ngnpope opened the pull request:
# I have made things!

Finishes off necessary changes to upgrade to mypy v1.16

## Related issues

Closes #2694, fixes #2696

sobolevn committed:
Update to mypy v1.16

Co-authored-by: Nick Pope <nick@nickpope.me.uk>

sobolevn committed:
Fix CI

sobolevn committed:
Fix `@cached_property` entries

ngnpope committed:
Remove `is_postgresql_14`, add `is_postgresql_17`

ngnpope committed:
Fix more `@cached_property` entries

These were not grouped with the other cases of this issue under the
following heading in the allowlist:

```
Ignore @cached_property error "cannot reconcile @property on stub with runtime object"
```

This fixes the following errors:

```
note: unused allowlist entry django.contrib.gis.db.backends.oracle.features.DatabaseFeatures.django_test_skips
note: unused allowlist entry django.contrib.gis.db.models.Lookup.output_field
note: unused allowlist entry django.db.backends.mysql.base.DatabaseWrapper.data_type_check_constraints
note: unused allowlist entry django.db.backends.mysql.base.DatabaseWrapper.display_name
note: unused allowlist entry django.db.backends.mysql.base.DatabaseWrapper.mysql_is_mariadb
note: unused allowlist entry django.db.backends.mysql.base.DatabaseWrapper.mysql_server_data
note: unused allowlist entry django.db.backends.mysql.base.DatabaseWrapper.mysql_server_info
note: unused allowlist entry django.db.backends.mysql.base.DatabaseWrapper.mysql_version
note: unused allowlist entry django.db.backends.mysql.base.DatabaseWrapper.sql_mode
note: unused allowlist entry django.db.models.Lookup.output_field
note: unused allowlist entry django.db.models.lookups.Lookup.output_field
note: unused allowlist entry django.db.models.sql.where.WhereNode.output_field
```

ngnpope committed:
Silence some `@cached_property` overrides for now

These are cached properties that override a plain attribute and so
stubtest complains with the following:

```
error: django.contrib.gis.db.backends.oracle.features.DatabaseFeatures.django_test_skips variable differs from runtime type functools.cached_property[Any]
Stub: in file /home/nick/Sources/_external/django-stubs/django-stubs/contrib/gis/db/backends/oracle/features.pyi:145
builtins.dict[builtins.str, builtins.set[builtins.str]]
Runtime:
<django.utils.functional.cached_property object at 0x7f628c1d8f50>

error: django.db.backends.oracle.features.DatabaseFeatures.django_test_skips variable differs from runtime type functools.cached_property[Any]
Stub: in file /home/nick/Sources/_external/django-stubs/django-stubs/db/backends/oracle/features.pyi:145
builtins.dict[builtins.str, builtins.set[builtins.str]]
Runtime:
<django.utils.functional.cached_property object at 0x7f628c1c5910>

error: django.db.backends.oracle.features.DatabaseFeatures.supports_frame_exclusion variable differs from runtime type functools.cached_property[Any]
Stub: in file /home/nick/Sources/_external/django-stubs/django-stubs/db/backends/oracle/features.pyi:102
builtins.bool
Runtime:
<django.utils.functional.cached_property object at 0x7f628c1c5b20>
```

There are many others of these that need looking at properly, so skip
them for now...

ngnpope committed:
Silence some more unfixable things

In the case of `Jinja2DivFormRenderer.__init__` - it will be gone come
Django 6.0 anyway when we can just remove this ignore.

ngnpope committed:
Fix issues due to variable override checking changes

Bisected to https://github.com/python/mypy/pull/18847

Some of these look valid, others might be due to bugs, e.g. duplicate
note output that might need to be fixed in mypy?

ngnpope committed:
Fix issues due to multiple inheritance checking changes

Bisected to https://github.com/python/mypy/pull/18876

It looks like this fixed a genuine problem! \o/

ngnpope commented on django-stubs/contrib/gis/db/backends/oracle/features.pyi:
```diff
@@ -1,14 +1,9 @@
     supports_geometry_field_unique_index: bool
     supports_perimeter_geodetic: bool
     supports_dwithin_distance_expr: bool
-    @cached_property
-    def django_test_skips(self) -> dict[str, Any]: ...  # type: ignore[override]
```
This wasn't quite right so I decided to ditch it as there are other cases of `django_test_skips` that are not handled yet. Will make sense to address them all at once.

ngnpope commented on scripts/stubtest/allowlist.txt:
```diff
@@ -268,3 +268,9 @@ django.contrib.gis.db.backends.base.features.BaseSpatialFeatures.unsupported_geo
 django.http.request.QueryDict.pop
+
+# Django incorrectly decorates this with `@classmethod`.
+django.test.selenium.SeleniumTestCase.__init_subclass__
```
`__init_subclass__` shouldn't be decorated with `@classmethod` even though it is one - it works implicitly. This is something that should be fixed in Django, but perhaps mypy is also unnecessarily picky here?

sobolevn replied:
yes, this is a `stubtest` bug. Can you please fix it upstread? I will be happy to review and merge your changes.

ngnpope replied:
Ah, ok. So I did a little digging and this isn't an issue in stubtest. While the explicit `@classmethod` is not required, it is not causing the issue - a lazy assumption I made - it's actually that there is the following in typeshed:

https://github.com/python/typeshed/blob/567b488fc28978642bfeb70b29f11453b0ff124c/stdlib/unittest/case.pyi#L289-L291

The issue is that Django's `SeleniumTestCase` eventually subclasses `TestCase` but the latter accepts `*args` while the former doesn't. So this is another issue that should be fixed in Django. I'll update the comment to more accurately reflect that.

ngnpope commented on scripts/stubtest/allowlist.txt:
```diff
@@ -268,3 +268,9 @@ django.contrib.gis.db.backends.base.features.BaseSpatialFeatures.unsupported_geo
+# Django incorrectly decorates this with `@classmethod`.
+django.test.selenium.SeleniumTestCase.__init_subclass__
+
+# Django incorrectly uses a signature of `(self, *args, **kwargs)` when it should be `(self)`.
+django.forms.renderers.Jinja2DivFormRenderer.__init__
```
This is just wrong in Django, but it will also be gone in Django 6.0 when we move to that in December.

ngnpope commented on scripts/stubtest/allowlist_todo.txt:
```diff
@@ -244,7 +244,6 @@ django.contrib.gis.db.models.JSONField.get_transform
 django.contrib.gis.db.models.Lookup.get_prep_lhs
 django.contrib.gis.db.models.Lookup.allowed_default
 django.contrib.gis.db.models.Lookup.lookup_name
-django.contrib.gis.db.models.Lookup.output_field
```
Most of these removals are just more cases of Django's `@cached_property` that weren't in the list that you'd already removed now that we have the workaround in mypy v1.16

ngnpope commented on tests/typecheck/contrib/admin/test_options.yml:
```diff
@@ -32,7 +35,7 @@
               ("group", {"fields": ("c",), "classes": ("a",), "description": "foo"}),
               (_("lazy"), {"fields": ["bar"]})
           ]
-          form = ModelForm
+          form = TestModelForm
```
This was incorrect - it should take a subclass of `ModelForm` - and it's now picked up 🎉

ngnpope commented on tests/typecheck/db/models/test_fields_choices.yml:
```diff
@@ -11,6 +11,10 @@
     main:4: note:         def __iter__(self) -> Iterator[Union[Tuple[Any, Any], Tuple[str, Iterable[Tuple[Any, Any]]]]]
     main:4: note:     Got:
     main:4: note:         def __iter__(self) -> Iterator[str]
+    main:4: note:     Expected:
+    main:4: note:         def __iter__(self) -> Iterator[Union[Tuple[Any, Any], Tuple[str, Iterable[Tuple[Any, Any]]]]]
+    main:4: note:     Got:
+    main:4: note:         def __iter__(self) -> Iterator[str]
```
This is annoying and perhaps a new bug in mypy v1.16? It doubles this note for some reason.

sobolevn replied:
Yes, looks like so. Please, report upsteam.

ngnpope replied:
Reported in https://github.com/python/mypy/issues/19240

ngnpope commented on tests/typecheck/db/models/test_fields_choices.yml:
```diff
@@ -25,6 +29,10 @@
     main:4: note:         def __iter__(self) -> Iterator[Union[Tuple[Any, Any], Tuple[str, Iterable[Tuple[Any, Any]]]]]
     main:4: note:     Got:
     main:4: note:         def __iter__(self) -> Iterator[str]
+    main:4: note:     Expected:
+    main:4: note:         def __iter__(self) -> Iterator[Union[Tuple[Any, Any], Tuple[str, Iterable[Tuple[Any, Any]]]]]
+    main:4: note:     Got:
+    main:4: note:         def __iter__(self) -> Iterator[str]
```
Ditto.

ngnpope commented on tests/typecheck/fields/test_related.yml:
```diff
@@ -779,10 +779,10 @@
         from myapp.models import User
-        reveal_type(User().parent)  # N: Revealed type is "myapp.models.User"
+        reveal_type(User().parent)  # N: Revealed type is "myapp.models.AbstractUser"
         reveal_type(User().parent_id)  # N: Revealed type is "builtins.int"
 
-        reveal_type(User().parent2)  # N: Revealed type is "Union[myapp.models.User, None]"
+        reveal_type(User().parent2)  # N: Revealed type is "Union[myapp.models.AbstractUser, None]"
```
I'm guessing these are more correct...

ngnpope commented on tests/typecheck/models/test_create.yml:
```diff
@@ -39,8 +39,6 @@
     main: |
         from myapp.models import Child4
         Child4.objects.create(name1='n1', name2='n2', value=1, value4=4)
-    out: |
-        myapp/models:9: error: Definition of "child1" in base class "Parent1" is incompatible with definition in base class "Parent2"  [misc]
```
Another thing that has been fixed 🎉

ngnpope commented on tests/typecheck/models/test_meta_options.yml:
```diff
@@ -66,7 +66,7 @@
                 abstract = 7  # E: Incompatible types in assignment (expression has type "int", base class "TypedModelMeta" defined the type as "bool")  [assignment]
                 verbose_name = ['test']  # E: Incompatible types in assignment (expression has type "List[str]", base class "TypedModelMeta" defined the type as "Union[str, _StrPromise]")  [assignment]
-                unique_together = {1: 2}  # E: Incompatible types in assignment (expression has type "Dict[int, int]", base class "TypedModelMeta" defined the type as "Union[Sequence[Sequence[str]], Sequence[str]]")  [assignment]
+                unique_together = {1: 2}  # E: Incompatible types in assignment (expression has type "Dict[int, int]", base class "TypedModelMeta" defined the type as "Sequence[Sequence[str]]")  [assignment]
```
This one is a bit weird. I'm presuming that `Sequence[Sequence[str]] | Sequence[str]` simplifies to just `Sequence[Sequence[str]]`... Maybe something we can simplify in the few places we do it:

```
ext/django_stubs_ext/db/models/__init__.py:39:        unique_together: ClassVar[Sequence[Sequence[str]] | Sequence[str]]
ext/django_stubs_ext/db/models/__init__.py:40:        index_together: ClassVar[Sequence[Sequence[str]] | Sequence[str]]  # Deprecated in Django 4.2
```

sobolevn replied:
Yes, let's do it. Good catch!

ngnpope replied:
Done.

sobolevn commented on .github/workflows/test.yml:
```diff
@@ -7,6 +7,10 @@ on:
   pull_request:
   workflow_dispatch:
 
+env:
```
It seems like we already have this on `main` branch https://github.com/typeddjango/django-stubs/blob/4a000e83ff0ce82e66c97178fc034ea939113a8a/.github/workflows/test.yml#L10

ngnpope replied:
See https://github.com/typeddjango/django-stubs/pull/2703#issuecomment-2943485106

sobolevn reviewed (commented):
THANKS A LOT! I am travelling right now, so I didn't have time to finish it. Amazing work.

One major question: it looks like there are some unrelated changed from the prev commit with `uv`

ngnpope committed:
Simplify types in `TypedModelMeta`

It seems that `Sequence[Sequence[str]] | Sequence[str]` simplifies to
`Sequence[Sequence[str]]` so we can remove this complexity. This more
simple annotation is also used for related values in
`django-stubs/db/backends/base/schema.pyi`.

ngnpope committed:
Update comment with correct reason

sobolevn reviewed (approved):
Thanks a lot! I will make a new release once I get back home.
